src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs

```diff
@@ -393,27 +394,125 @@ private Task StartAsyncCore(TransferFormat transferFormat)
             }
         }
 
-        private static async Task<NegotiationResponse> ParseNegotiateResponse(HttpResponseMessage response, ILogger logger)
+        private static async Task<NegotiationResponse> ParseNegotiateResponse(HttpResponseMessage response)
         {
-            NegotiationResponse negotiationResponse;
-            using (var reader = new JsonTextReader(new StreamReader(await response.Content.ReadAsStreamAsync())))
+            try
             {
-                try
-                {
-                    negotiationResponse = new JsonSerializer().Deserialize<NegotiationResponse>(reader);
-                }
-                catch (Exception ex)
+                using (var reader = new JsonTextReader(new StreamReader(await response.Content.ReadAsStreamAsync())))
                 {
-                    throw new FormatException("Invalid negotiation response received.", ex);
+                    JsonUtils.CheckRead(reader);
+                    JsonUtils.EnsureObjectStart(reader);
+
+                    string connectionId = null;
+                    List<AvailableTransport> availableTransports = null;
+
+                    var completed = false;
+                    do
+                    {
+                        switch (reader.TokenType)
+                        {
+                            case JsonToken.PropertyName:
+                                string memberName = reader.Value.ToString();
+
+                                switch (memberName)
+                                {
+                                    case "connectionId":
+                                        connectionId = JsonUtils.ReadAsString(reader, "connectionId");
+                                        break;
+                                    case "availableTransports":
+                                        JsonUtils.CheckRead(reader);
+                                        JsonUtils.EnsureArrayStart(reader);
+
+                                        availableTransports = new List<AvailableTransport>();
```

Comment:
Create with capacity 3

Reply:
The default capacity of a list is 4. It would be a very small optimization.